cpp/util/ManagedArray.h

```diff
@@ -0,0 +1,305 @@
+ *  ManagedArray and its underlying data between compute calls.
+ *
+ *  There are two specific scenarios when ManagedArrays are invalidated.
+ *      1. ManagedArrays in state 2 will be invalidated if the data source
+ *         array is deleted.
+ *      2. ManagedArrays in state 3 will be invalidated if the data source is
+ *         deleted or changes in size.
+ *
+ *  No attempt is made to protect against these failures. The user is
+ *  responsible for avoiding these cases.
+ */
+template<typename T> class ManagedArray
+{
+public:
+    //! Default constructor.
+    /*! For safety, these arrays must be constructed with the manage argument
+     *  set to true in order to create an empty instance that manages its own
+     *  memory. By default, constructed arrays do not manage their own memory to
+     *  avoid accidental memory leaks.
+     *
+     *  \param manage Whether or not this instance should manage its own memory
+     *                (Default: false).
+     */
+    ManagedArray(bool manage=false) : m_size(nullptr), m_data(nullptr), m_managed(manage), m_external_ptr(false) {}
+
+    //! Constructor with specific size for thread local arrays
+    /*! When using this constructor, the class automatically manages its own
+     *  memory since it is allocating it.
+     *
+     *  \param size Size of the array to allocate.
+     */
+    ManagedArray(unsigned int size) : m_managed(true), m_external_ptr(false)
+    {
+        m_size = new unsigned int(size);
+        m_data = new T[size];
+        reset();
+    }
+
+    //! Construct object from existing array.
+    /*! \param T* Pointer to existing data.
+     *  \param size Size of the array to allocate.
+     */
+    ManagedArray(T* array, unsigned int size) : m_managed(false), m_external_ptr(true)
+    {
+        m_size = new unsigned int(size);
+        m_data = array;
+    }
+
+    //! Copy constructor.
+    /*! This constructor is required to ensure that the original object always
+     *  owns its own memory. However ManagedArrays pointing to external data will always store
+     *
+     *  \param other ManagedArray instance to copy.
+     */
+    ManagedArray(const ManagedArray &other) : m_data(other.m_data), m_managed(false), m_external_ptr(other.m_external_ptr)
+    {
+        if (m_external_ptr)
+        {
+            m_size = new unsigned int(*other.m_size);
+        }
+        else
+        {
+            m_size = other.m_size;
+        }
+    }
+
+    //! Copy assignment.
+    /*! Similar to the copy constructor, this operator must be defined to
+     *  ensure that the original object always owns its own memory.
+     *
+     *  \param first ManagedArray instance to copy.
+     */
+    ManagedArray& operator= (const ManagedArray &other)
+    {
+        if (m_managed)
+        {
+            throw std::runtime_error("You cannot assign to a ManagedArray that is currently managing its own memory.");
```

Comment:
Maybe you had a different use case in mind, but why not? Can't you just deallocate the memory this object owns and adjust the flags appropriately?

Reply:
The reason I forbade this is to minimize the possibility of ManagedArrays lying around that pointed to invalidated memory (because the ManagedArray owning the memory was deallocated in this fashion). But I agree that it's perhaps a confusing restriction. If we switch to using smart pointers as discussed in the other comments, this issue goes away.